Install examples

permcody opened the pull request:
This is a first rough take at supporting installing more than just the "tests" directory. I think there are many more improvements that could be added, but this may be good enough to get us started considering how critical this is. I'm happy to attempt to address 😉 whatever I can, but here are some thoughts:

- Really don't like the shell loop in the Makefile. This needs to be converted to a proper Make dependency for each installable directory instead
- Not my bug, but I noticed that you cannot run `make install` before running `make`. We are missing a dependency in there. It appears that the doc installer kicks off before the binary is built. Shouldn't be too bad to track this down.
- Probably needs some more documentation. This will be used by binary users who have no idea what may be available to copy so perhaps we need a way to help the user know what the valid arguments are for a given application. I don't want to bake this into the binary and the only place it exists right now is the main Makefile which we don't parse (and probably don't want to). Need to think more about what to do to help the user. 


This PR shouldn't break any of the install targets as I have not yet removed the existing `--copy-tests` and `--tests` CLI options. Those two default to the old behavior. To get an application or module to install more than one directory you need to specify the **relative paths with respect to the root of your repository** in your `Makefile`:

e.g.
`INSTALLABLE_DIRS := test/tests examples problems`

Note: Do not quote the above line, it'll turn it into a string which doesn't allow the shell loop to loop over it. This goes back to the first improvement that could be implemented (later?), in turning each installed directory into a Makefile target instead of just looping over a variable. If the above line doesn't exist then we default to just `test/tests`.

permcody committed:
Makefile changes to support multiple installable directories

refs #19022

permcody committed:
Generalizing the install and run logic to handle more than just "tests"

refs #19022

Apply suggestions from code review

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/962551/) on 8bb357a wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20348/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/962559/) on 8bb357a wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>271672</th>
      <th colspan=4>#20348 8bb357</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.33%</td>
      <td>82.33%</td>
      <td>-0.00%</td>
      <td>44.64%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>73811</td>
      <td>73825</td>
      <td>+14</td>
      <td>25</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15839</td>
      <td>15846</td>
      <td>+7</td>
      <td>31</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20348/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20348/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20348/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 44.64% is less than the suggested 90.0%



This comment will be updated on new commits.

loganharbour reviewed (commented):
We should probably tests for this, even for the errors.

permcody committed:
Fixing up tests

Getting rid of assumptions in tests that rely on the file structure further up than the copy point. This
includes the relative location of the executable.

Also removing the empty executable option. While this isn't an unreasonable test, we don't have a good
capability today to run a test that relies on a relative path outside of the installed tree. We may re-add
this capability at a later date. By switching all the useability tests back to RunApp we can't control the
CLI arguments that are passed to the application, invalidating the need for that test.

refs #19022

moosebuild commented:
Job [Test install](https://civet.inl.gov/job/975298/) on a13bb7c : invalidated by @permcody

permcody committed:
Installable Dirs variable move

INSTALLABLE_DIRS default initializatin needs to be moved before the stack operation or it
will always be populated.

moosebuild commented:
Job [Min gcc](https://civet.inl.gov/job/976083/) on af4d19b : invalidated by @permcody

moosebuild commented:
Job [Min gcc](https://civet.inl.gov/job/976083/) on af4d19b : invalidated by @permcody

permcody commented:
I think this is good enough for a round one. I plan to look into testing the install and run calls. I'm pretty sure I can come up with a way to do that. Additionally, I have some Makefile improvements that I'd like to get in here too. Neither of these should hold us up from merging though as we can start updating recipes and deprecating the old syntax after this is merged and then I can clean that up too.

Thanks!

loganharbour commented on framework/src/base/MooseApp.C:
```diff
@@ -1371,105 +1380,148 @@ MooseApp::run()
+      mooseError(
+          "The directory \"./",
+          dst_dir,
+          "\" already exists.\nTo update/recopy the contents of this direcotry, rename (\"mv ",
```
```suggestion
          "\" already exists.\nTo update/recopy the contents of this directory, rename (\"mv ",
```

loganharbour commented on framework/src/base/MooseApp.C:
```diff
@@ -1371,105 +1380,148 @@ MooseApp::run()
+    // Get command line argument following --run on command line
+    dir_to_run = getParam<std::string>("run");
+    run_param_valid = true;
+    skip_args = 3;
```
is run always the first arg?

permcody replied:
Well it should be for running examples since it wouldn't make sense to send commands to MOOSE and run the TestHarness at the same time. If people do that now, and list run later, things are not going to work correctly. If they have MOOSE options after the run argument, they'll get passed to the TestHarness and that'll likely throw an error. This is all terrible logic that I am admittedly propagating. It can all be cleaned up.

permcody committed:
Logic cleanup, better error messages

permcody committed:
Build system testing and documentation

refs #19022

loganharbour commented on modules/doc/content/application_development/build_system.md:
```diff
@@ -208,3 +208,49 @@ when building MOOSE or a MOOSE-based application.
+
+### Make Install
+
+MOOSE's build system has the ability to create installed binaries suitable for use with the conda package manager, or installation on a shared computing resource such as Sawtooth. The `install` target will copy binary and required libraries to a file tree based on the `PREFIX` you configured with.
```
```suggestion
MOOSE's build system has the ability to create installed binaries suitable for use with the conda package manager, or installation on a shared computing resource such as Sawtooth. The `install` target will copy binary and required libraries to a file tree based on the prefix you configured with (set by the variable `PREFIX` or the `--prefix` argument.
```

loganharbour commented on modules/doc/content/application_development/build_system.md:
````diff
@@ -208,3 +208,49 @@ when building MOOSE or a MOOSE-based application.
+$ make install
+```
+
+Additionally each application may wish to make an one or more file structures availabel for end users to install into their own environment. This allows for end-users to install and run the application test suite, examples, tutorials, or other inputs of interest. To customize what is available for installation, the application developer will define the list of directories in their application mak ...
````
```suggestion
Additionally each application may wish to make an one or more file structures available for end users to install into their own environment. This allows for end-users to install and run the application test suite, examples, tutorials, or other inputs of interest. To customize what is available for installation, the application developer will define the list of directories in their application make file. For example:
```

loganharbour commented on modules/doc/content/application_development/build_system.md:
````diff
@@ -208,3 +208,49 @@ when building MOOSE or a MOOSE-based application.
+```
+$ cd <user writable location>
+$ export PATH=$PATH:<prefix>/bin  # See instructions for HPC computers below
+$ bison-opt --copy-inputs <dir>
+$ bison-opt --run <dir>
+
+```
+
+Where `<dir>` is one of the installable directories as defined by the application developers (i.e. `tests`, `examples`, `tutorials`, etc.)
````
I think `<dir>` is confusing here. On first inspection I wondered why the `cd` was necessary and thought `<dir>` would be where it was copied. Obviously after reading more this went away, just a thought.

loganharbour commented on modules/doc/content/application_development/build_system.md:
```diff
@@ -208,3 +208,49 @@ when building MOOSE or a MOOSE-based application.
+
+Where `<dir>` is one of the installable directories as defined by the application developers (i.e. `tests`, `examples`, `tutorials`, etc.)
+
+When using INL Cluster systems to run your input, you will load a module that will set your path correctly
```
```suggestion
When using INL HPC systems to run your input, you will load a module that will set your path correctly
```

loganharbour commented on modules/doc/content/application_development/build_system.md:
````diff
@@ -208,3 +208,49 @@ when building MOOSE or a MOOSE-based application.
+Example:
+
+```
+$ module load use.moose <app name> # e.g BISON
````
```suggestion
$ module load use.moose <app name> # e.g bison
```

loganharbour commented on modules/doc/content/application_development/build_system.md:
```diff
@@ -208,3 +208,49 @@ when building MOOSE or a MOOSE-based application.
 - `framework/conf_var.mk.in` - This file is where you put your new expansion expressions. Typically these
   are in the form of "@VARAIBLE@".
+
+### Make Install
```
Would be nice if we could set this up as a template so that application documentation could include it

loganharbour commented on test/tests/make_install/tests:
```diff
@@ -0,0 +1,49 @@
+      input = 'DUMMY'
+      cli_args = "--copy-inputs tests"
+      expect_out = 'Directory successfully copied'
+      prereq = 'setup_fake_test_structure'
```
I don't think the prereqs work when you have nested tests, but I can't remember

permcody committed:
Adding TestHarness caveat for "installed"

permcody committed:
Make copy and run commands more robust for parallel launches

refs #19022

permcody committed:
Adding an option to prevent the TestHarness from appending addition CLI options

permcody commented:
@loganharbour - should be ready to go!

permcody committed:
Drop the Makefile Loop

refs #19022

permcody committed:
Better default for INSTALLABLE_DIRS

refs #19022

permcody committed:
Feel like I'm getting into hacking territory

It turns out that there are some baked in assumptions of where the App is located
relative to the testroot. There are some nasty hacks to locate it in a normal
installation location, but this doesn't work if you install elsewhere (like we have
to for the test). In order to get the TestHarness to behave, it's easiest to just
link the executable to the fake test hierarchy. This part isn't a hack ;)

Next Problem, the TestHarness currently checks for the working directory to exist
in the Runnable phase, which occurs asyncronously to the scheduled launch of the
test. This is a no go, since we are trying to copy the directory structure in the test
so it won't exist for latter tests. So... I moved the working directory check to
the place where we actually launch the test. This would be OK, but apparently there
are other assumptions baked into the Job class. I had to start/stop the timer to
avoid an index out of range.

Now the tests are running properly.

refs #19022

YaqiWang commented on modules/doc/content/application_development/build_system.md:
````diff
@@ -208,3 +208,49 @@ when building MOOSE or a MOOSE-based application.
+```
+$ cd moose/framework
+$ ./configure --prefix=<installation path>
+$ make
````
Missing `cd ../..` back to the app directory.

permcody replied:
Thanks Yaqi - I will be working on this more as I deprecate the old options and tune this up to meet the needs of Griffin and other requirements that I may have overlooked. This will be useful for me to improve the documentation.

YaqiWang commented on modules/doc/content/application_development/build_system.md:
````diff
@@ -208,3 +208,49 @@ when building MOOSE or a MOOSE-based application.
+Additionally each application may wish to make an one or more file structures available for end users to install into their own environment. This allows for end-users to install and run the application test suite, examples, tutorials, or other inputs of interest. To customize what is available for installation, the application developer will define the list of directories in their application mak ...
+
+```
+INSTALLABLE_DIRS := test/tests examples
````
What does this mean? Does this copy `test/tests` into `<installation path>/examples`?

YaqiWang replied:
And `examples` is considered as a type? How if I have something like
```
INSTALLABLE_DIRS += test/foo examples/foo
```
?

permcody replied:
You may be overthinking this. This is simply the list of directories that will be installable. It's not a "To" or "From", or a "type". 
In the above example, users would be able to install "tests" and "examples". There is one point of clarification though and I can see now that there is a potential deficiency in the way I'm currently handling it. If you supply a path with slashes in it, I'm currently only taking the last directory as the directory a user will type in the `--copy-inputs` command (so "tests" or "examples"). If you have more complicated structures like "tests/installable examples/installable problems/installable", it's not going to work as it's coded now. 

However, I don't want people to end up with "test_tests" or other complicated things to remember so I didn't try to automatically flatten these structures... I might leave it this way for right now and if needed code up a way for developers to give their installed directories.

YaqiWang commented on modules/doc/content/application_development/build_system.md:
```diff
@@ -208,3 +208,49 @@ when building MOOSE or a MOOSE-based application.
+
+Where `<installable input type>` is one of the installable directories as defined by the application developers (i.e. `tests`, `examples`, `tutorials`, etc.)
+
+When using INL HPC systems to run your input, you will load a module that will set your path correctly
```
will load => will need to load? Missing period sign at the end.

YaqiWang commented on modules/doc/content/application_development/build_system.md:
```diff
@@ -208,3 +208,49 @@ when building MOOSE or a MOOSE-based application.
+$ cd <user writable location>
+$ export PATH=$PATH:<prefix>/bin  # See instructions for HPC computers below
+$ bison-opt --copy-inputs <installable input type>  # e.g. tests, examples, etc.
+$ bison-opt --run <installable input type>
```
What is the difference between this and `--tests` option?

permcody replied:
Nothing... `--tests` was too specific so it's deprecated and getting removed very soon.

YaqiWang commented on modules/doc/content/application_development/build_system.md:
````diff
@@ -208,3 +208,49 @@ when building MOOSE or a MOOSE-based application.
+
+MOOSE's build system has the ability to create installed binaries suitable for use with the conda package manager, or installation on a shared computing resource such as Sawtooth. The `install` target will copy binary and required libraries to a file tree based on the prefix you configured with (set by the variable `PREFIX` or the `--prefix` argument.
+
+```
````
```language=bash might make this look better. I did not check the html page though.

YaqiWang commented on modules/doc/content/application_development/build_system.md:
````diff
@@ -208,3 +208,49 @@ when building MOOSE or a MOOSE-based application.
+MOOSE's build system has the ability to create installed binaries suitable for use with the conda package manager, or installation on a shared computing resource such as Sawtooth. The `install` target will copy binary and required libraries to a file tree based on the prefix you configured with (set by the variable `PREFIX` or the `--prefix` argument.
+
+```
+$ cd moose/framework
````
I think `configure` is under `moose` but not `moose/framework`.

permcody commented:
Thanks for all the comments. I'll write up better documentation to answer
all this.

On Mon, Mar 14, 2022 at 1:03 PM Yaqi ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In modules/doc/content/application_development/build_system.md
> <https://github.com/idaholab/moose/pull/20348#discussion_r826287210>:
>
> > @@ -208,3 +208,49 @@ when building MOOSE or a MOOSE-based application.
>
>  - `framework/conf_var.mk.in` - This file is where you put your new expansion expressions. Typically these
>    are in the form of ***@***.***@".
> +
> +### Make Install
> +
> +MOOSE's build system has the ability to create installed binaries suitable for use with the conda package manager, or installation on a shared computing resource such as Sawtooth. The `install` target will copy binary and required libraries to a file tree based on the prefix you configured with (set by the variable `PREFIX` or the `--prefix` argument.
> +
> +```
> +$ cd moose/framework
>
> I think configure is under moose but not moose/framework.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/20348#pullrequestreview-909267838>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAXFOIAQA6ZMUGHWKT6MXZLU76EOJANCNFSM5OZIMSKA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
